subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -232,54 +235,56 @@ private void defineConfigurationsForSourceSet(SourceSet sourceSet, Configuration
         annotationProcessorConfiguration.setDescription("Annotation processors and their dependencies for " + sourceSetName + ".");
-        annotationProcessorConfiguration.setCanBeConsumed(false);
-        annotationProcessorConfiguration.setCanBeResolved(true);
-
         jvmPluginServices.configureAsRuntimeClasspath(annotationProcessorConfiguration);
 
-        Configuration runtimeOnlyConfiguration = configurations.maybeCreate(runtimeOnlyConfigurationName);
+        warnIfConfigurationAlreadyExists(configurations, runtimeOnlyConfigurationName);
+        Configuration runtimeOnlyConfiguration = configurations.maybeCreateWithRole(runtimeOnlyConfigurationName, ConfigurationRoles.INTENDED_BUCKET, false, true);
         runtimeOnlyConfiguration.setVisible(false);
-        runtimeOnlyConfiguration.setCanBeConsumed(false);
-        runtimeOnlyConfiguration.setCanBeResolved(false);
         runtimeOnlyConfiguration.setDescription("Runtime only dependencies for " + sourceSetName + ".");
 
-        Configuration runtimeClasspathConfiguration = configurations.maybeCreate(runtimeClasspathConfigurationName);
+        // TODO: The jmh plugin prevents asserting this role upon creation; if it has been applied, then it will pre-create a runtime classpath configuration
+        // for the jmg sourceset, which is both resolvable and declarable against, so this would fail if we assert the role's usage here.
+        warnIfConfigurationAlreadyExists(configurations, runtimeClasspathConfigurationName);
+        Configuration runtimeClasspathConfiguration = configurations.maybeCreateWithRole(runtimeClasspathConfigurationName, ConfigurationRoles.INTENDED_RESOLVABLE, false, false);
         runtimeClasspathConfiguration.setVisible(false);
-        runtimeClasspathConfiguration.setCanBeConsumed(false);
-        runtimeClasspathConfiguration.setCanBeResolved(true);
         runtimeClasspathConfiguration.setDescription("Runtime classpath of " + sourceSetName + ".");
         runtimeClasspathConfiguration.extendsFrom(runtimeOnlyConfiguration, implementationConfiguration);
         jvmPluginServices.configureAsRuntimeClasspath(runtimeClasspathConfiguration);
 
         sourceSet.setCompileClasspath(compileClasspathConfiguration);
         sourceSet.setRuntimeClasspath(sourceSet.getOutput().plus(runtimeClasspathConfiguration));
         sourceSet.setAnnotationProcessorPath(annotationProcessorConfiguration);
+    }
 
-        compileClasspathConfiguration.setCanBeDeclaredAgainst(false);
-        ((ConfigurationInternal) runtimeClasspathConfiguration).setCanBeDeclaredAgainst(false);
+    private void warnIfConfigurationAlreadyExists(ConfigurationContainer configurations, String configurationName) {
+        if (configurations.findByName(configurationName) != null) {
+            DeprecationLogger.deprecateBehaviour("The configuration " + configurationName + " was created explicitly. This configuration name is reserved for creation by Gradle.")
```

Comment:
I disagree here.  It's kind of been implied for a long time that you shouldn't create configurations with these names.  If you are doing this, you can deal with the extra nagging.

Reply:
I'm not sure if we have this _written down_ anywhere, but it is a goal of ours to not have new deprecation warnings in major releases. Just a week ago, [we reverted a PR](https://github.com/gradle/gradle/pull/22900) just to avoid new deprecation warnings in 8.0. The intention is that if you have a clean build in the final minor version (7.6) without deprecation warnings, you should be safe to upgrade to the next major (8.0) without additional warnings. 

We've made an effort to avoid adding new deprecations in 8.0 so far, why go against that now?

Reply:
I hear you...but even though this might not have a formal deprecation warning or documentation, I have a hard time believing anyone creating an `implementation` configuration ahead of the Java plugin doesn't already know they're doing something a little naughty.

I don't feel too strongly about this though.  What does the rest of the team think?

Reply:
Resolving since this is merging into 8.1 and deprecation warnings are okay